validator/keymanager/remote-web3signer/keymanager.go

```diff
@@ -38,54 +53,327 @@ type SetupConfig struct {
+		ppk = cfg.ProvidedPublicKeys
+	}
+
+	// use a map to remove duplicates
+	flagLoadedKeys := make(map[string][48]byte)
+
+	// Populate the map with existing keys
+	for _, key := range ppk {
+		decodedKey, err := hexutil.Decode(key)
+		if err != nil {
+			return nil, errors.Wrap(err, "could not decode public key for remote signer")
+		}
+		if len(decodedKey) != fieldparams.BLSPubkeyLength {
+			return nil, fmt.Errorf("invalid public key length: expected %d, got %d", fieldparams.BLSPubkeyLength, len(decodedKey))
+		}
+		flagLoadedKeys[key] = bytesutil.ToBytes48(decodedKey)
+	}
+	km.flagLoadedKeysMap = flagLoadedKeys
+
+	// load from file
+	if keyFileExists {
+		_, fileKeys, err := km.readKeyFile()
+		if err != nil {
+			return nil, errors.Wrap(err, "could not read key file")
 		}
-		// makes sure that if the public keys are deleted the validator does not call URL again.
-		km.publicKeysUrlCalled = true
-		km.providedPublicKeys = providedPublicKeys
+		maps.Copy(fileKeys, flagLoadedKeys)
+		km.lock.Lock()
+		km.providedPublicKeys = maps.Values(fileKeys)
+		km.lock.Unlock()
+		// create a file watcher
+		go func() {
+			err = km.refreshRemoteKeysFromFileChangesWithRetry(ctx, retryDelay)
+			if err != nil {
+				log.WithError(err).Error("could not refresh remote keys from file changes")
+			}
+		}()
+	} else {
+		km.lock.Lock()
+		km.providedPublicKeys = maps.Values(flagLoadedKeys)
+		km.lock.Unlock()
+	}
+
+	return km, nil
+}
+
+func checkFileExists(path string, retriesRemaining int, retryDelay time.Duration) (bool, error) {
+	rr := retriesRemaining // make a copy to not alter the original
+	if rr == 0 {
+		return false, errors.New("file check retries remaining exceeded")
 	}
+
+	// Check if the directory and file exists, if not create it
+	keyFileExists, err := file.Exists(path, file.Regular)
+	if err != nil {
+		return false, errors.Wrapf(err, "could not check if remote signer persistent keys exists in %s", path)
+	}
+
+	if !keyFileExists {
+		rr--
+		log.WithFields(logrus.Fields{"path": path, "retries_remaining": rr, "retry_delay": retryDelay}).Warnf("Failed to find the key file. Retrying until the file is detected ...")
+		time.Sleep(retryDelay)
+		return checkFileExists(path, rr, retryDelay)
+	}
+	return true, nil
+}
+
+func (km *Keymanager) refreshRemoteKeysFromFileChangesWithRetry(ctx context.Context, retryDelay time.Duration) error {
+	if km.retriesRemaining == 0 {
+		return errors.New("file check retries remaining exceeded")
+	}
+	err := km.refreshRemoteKeysFromFileChanges(ctx)
+	if err != nil {
+		km.updatePublicKeys(make([][48]byte, 0)) // update the keys to empty as the file was never loaded correctly
+		km.retriesRemaining--
+		log.WithError(err).Debug("could not refresh remote keys from file changes")
+		log.WithFields(logrus.Fields{"path": km.keyFilePath, "retries_remaining": km.retriesRemaining, "retry_delay": retryDelay}).Warnf("Key file does not exist. Please create a new one. Retrying ...")
+		time.Sleep(retryDelay)
```

Comment:
What about checking if the context is cancelled before the sleep?

Reply:
Is this comment addressed?

Reply:
this was addressed, github is not updating the file correctly... you can see it here https://github.com/prysmaticlabs/prysm/pull/13682/commits/d447fcf6fc0a709e67827a36f3856d620f44fe67